adapters/yandex/yandex.go

```diff
@@ -0,0 +1,311 @@
+	currency := getCurrency(requestData)
+
+	for i := range requestData.Imp {
+		imp := requestData.Imp[i]
+		var yandexExt openrtb_ext.ExtImpYandex
+
+		if err := getYandexImpExt(imp, &yandexExt); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		if err := modifyImp(&imp); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		var resolvedUrl, err = a.resolveUrl(yandexExt, referer, currency)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		splittedRequestData := splitRequestDataByImp(requestData, imp)
+
+		requestBody, err := json.Marshal(splittedRequestData)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		requests = append(requests, &adapters.RequestData{
+			Method:  "POST",
+			Uri:     resolvedUrl,
+			Body:    requestBody,
+			Headers: getHeaders(&splittedRequestData),
+		})
+	}
+
+	return requests, errors
+}
+
+func getHeaders(request *openrtb2.BidRequest) http.Header {
+	headers := http.Header{}
+
+	if request.Device != nil {
+		addNonEmptyHeaders(&headers, map[string]string{
+			"Accept-Language": request.Device.Language,
+			"User-Agent":      request.Device.UA,
+			"X-Forwarded-For": request.Device.IP,
+			"X-Real-Ip":       request.Device.IP,
+			"Content-Type":    "application/json;charset=utf-8",
+			"Accept":          "application/json",
+		})
+	}
+
+	return headers
+}
+
+func addNonEmptyHeaders(headers *http.Header, headerValues map[string]string) {
+	for key, value := range headerValues {
+		if len(value) > 0 {
+			headers.Add(key, value)
+		}
+	}
+}
+
+// Request is in shared memory, so we have to make a shallow copy for further modification (imp is already a shallow copy)
+func splitRequestDataByImp(request *openrtb2.BidRequest, imp openrtb2.Imp) openrtb2.BidRequest {
+	requestCopy := *request
+	requestCopy.Imp = []openrtb2.Imp{imp}
+
+	return requestCopy
+}
+
+func getYandexImpExt(imp openrtb2.Imp, yandexExt *openrtb_ext.ExtImpYandex) error {
+	var ext adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &ext); err != nil {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("imp %s: unable to unmarshal ext", imp.ID),
+		}
+	}
+
+	if err := json.Unmarshal(ext.Bidder, yandexExt); err != nil {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("imp %s: unable to unmarshal ext.bidder", imp.ID),
+		}
+	}
+
+	return nil
+}
+
+func modifyImp(imp *openrtb2.Imp) error {
+	if imp.Banner != nil {
+		banner, err := modifyBanner(*imp.Banner)
+		if banner != nil {
+			imp.Banner = banner
+		}
+		return err
+	}
+
+	if imp.Native != nil {
+		return nil
+	}
+
+	return &errortypes.BadInput{
+		Message: fmt.Sprintf("Unsupported format. Yandex only supports banner and native types. Ignoring imp id #%s", imp.ID),
+	}
+}
+
+func modifyBanner(banner openrtb2.Banner) (*openrtb2.Banner, error) {
+	height := banner.H
+	width := banner.W
+	format := banner.Format
+
+	if width == nil || height == nil || *width == 0 || *height == 0 {
```

Comment:
`height, width` are only used in if check added in line 163.  Code can be modified as below:

```
format := banner.Format
       if banner.W == nil || banner.H == nil || *banner.W == 0 || *banner.H == 0 {
            if len(format) == 0 {
	        return nil, &errortypes.BadInput{Message: "Invalid size provided for Banner"}
             }

           firstFormat := format[0]
           banner.H = &firstFormat.H
           banner.W = &firstFormat.W
       }
      return &banner, nil
}


```